app/src/main/java/com/example/tictactoe/model/TicTacToeViewModel.kt

```diff
@@ -32,4 +56,172 @@ class TicTacToeViewModel : ViewModel() {
+
+        // Column 2
+        if (testRow(GameBoardPart.COLUMN_MIDDLE.part, GameBoardPart.BOX_CENTER.part)) return true
+
+        // Column 3
+        if (testRow(
+                GameBoardPart.COLUMN_RIGHT.part,
+                GameBoardPart.BOX_BOTTOM_RIGHT.part
+            )
+        ) return true
+
+        /* Horizontal checks for the win:
+               If the set of a specified row does not contain an empty space,
+               and is equal to the set of any single space within it,
+               it must be three in a row */
+        // Row 1
+        if (testRow(GameBoardPart.ROW_TOP.part, GameBoardPart.BOX_TOP_LEFT.part)) return true
+
+        // Row 2
+        if (testRow(GameBoardPart.ROW_MIDDLE.part, GameBoardPart.BOX_CENTER.part)) return true
+
+        // Row 3
+        if (testRow(GameBoardPart.ROW_BOTTOM.part, GameBoardPart.BOX_BOTTOM_RIGHT.part)) return true
+
+        /* Diagonal checks for the win:
+               If the set of a specified diagonal row does not contain an empty space,
+               and is equal to the set of any single space within it,
+               it must be three in a row */
+        // Top left to bottom right
+        if (testRow(
+                GameBoardPart.DIAGONAL_TOP_LEFT_TO_BOTTOM_RIGHT.part,
+                GameBoardPart.BOX_CENTER.part
+            )
+        ) return true
+
+        // Top right to bottom left
+        if (testRow(
+                GameBoardPart.DIAGONAL_TOP_RIGHT_TO_BOTTOM_LEFT.part,
+                GameBoardPart.BOX_CENTER.part
+            )
+        ) return true
+
+        return false
+    }
+
+    /**
+     * Tests given row to see if it is three in a row for a player.
+     * If it is, it will return true and announce said player.
+     *
+     * @param row The row being tested for a win
+     *
+     * @param box A box within that row that can be used to compare sets for a win
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun testRow(row: Set<String>, box: Set<String>): Boolean {
+        return if (!(row.contains(""))) {
+            if (row == box) {
+                // Convert set of single box to list to allow access to contained string
+                announceWinner(box.toList()[0])
+                return true
+            }
+            false
+        } else {
+            false
+        }
+    }
+
+    /**
+     * Retrieves necessary image to place in clicked space.
+     *
+     * @param boardValue Value of space in game board
+     *
+     * @return Resource Id for the current player's symbol
+     */
+    fun getIcon(boardValue: String?): Int {
+        return try {
+            when (boardValue) {
+                "X" -> R.drawable.icon_x
+                "O" -> R.drawable.icon_o
+                else -> throw IllegalArgumentException("Must pass valid symbol to getIcon")
+            }
+        } catch (e: IllegalArgumentException) {
+            Log.e(TAG, e.message.toString())
```

Comment:
what you'll find a lot of people do in this scenario is just use the class name `this::class.java.simpleName` is the tag - and what you'll find popular logging libraries like Timber do you for out of the box - makes it less brittle to refactors

Reply:
Thank you! I've heard of timber but hadn't gotten the chance to use it yet. I did a little research and switched it out for my own custom tag.